lambda_function.py

```diff
@@ -293,64 +189,203 @@ def gather_table_stats(cursor, cluster):
+                    'MetricName': command['name'],
+                    'Dimensions': [
+                        {'Name': self.environment, 'Value': self.cluster}
+                    ],
+                    'Timestamp': t,
+                    'Value': interval,
+                    'Unit': 'Milliseconds'
+                })
+
+        return output_metrics
+
+    @staticmethod
+    def run_command(cursor, statement):
+        logger.debug("Running Statement: %s" % statement)
+
+        t = datetime.datetime.now()
+        try:
+            cursor.execute(statement)
+        except pg8000.core.ProgrammingError as e:
+            logger.error('Error executing statement %s: %s', statement, e)
+        interval = (datetime.datetime.now() - t).microseconds / 1000
+
+        return interval
+
+    def gather_service_class_stats(self, cursor):
+        metrics = []
+        poll_ts = datetime.datetime.utcnow()
+        _ = self.run_command(cursor, '''SELECT service_class, num_queued_queries, num_executing_queries 
+                                            FROM stv_wlm_service_class_state w WHERE w.service_class >= 6 ORDER BY 1
+                                    ''')
+        service_class_info = cursor.fetchall()
+
+        for service_class in service_class_info:
+            queued_metric = {'MetricName': 'ServiceClass%s-Queued' % service_class[0],
+                             'Dimensions': [{'Name': self.environment, 'Value': self.cluster}],
+                             'Timestamp': poll_ts,
+                             'Value': service_class[1]}
+            metrics.append(queued_metric.copy())
+
+            executing_metric = {'MetricName': 'ServiceClass%s-Executing' % service_class[0],
+                                'Dimensions': [{'Name': self.environment, 'Value': self.cluster}],
+                                'Timestamp': poll_ts,
+                                'Value': service_class[2]}
+            metrics.append(executing_metric.copy())
+
+        return metrics
+
+
+def configure():
+    reporter = Reporter()
+
+    if reporter.debug:
+        logger.setLevel(logging.DEBUG)
+        logger.debug("DEBUG mode set.")
+    else:
+        logger.setLevel(logging.INFO)
+
+    for k, v in attr.asdict(reporter).items():
+        if not v:
+            logger.warning("Parameter {} is unset".format(k))
+
+    return reporter
+
+
 def lambda_handler(event, context):
+    reporter = configure()
+
+    cw = boto3.client('cloudwatch', region_name=reporter.region)
+
+    pg8000.paramstyle = "qmark"
+
     try:
-        if debug:
-            print('Connecting to Redshift: %s' % host)
-        conn = pg8000.connect(database=database, user=user, password=password, host=host, port=port, ssl=ssl)
+        kms = boto3.client('kms', region_name=reporter.region)
+        # Check if decryption is possible
+        password = kms.decrypt(CiphertextBlob=base64.b64decode(reporter.encrypted_password))['Plaintext']
```

Comment:
Guessing this comes out of the lambda env with encrypted KMS helper?

Reply:
As long as you have `.aws/credentials` file in one of the appropriate places, or are on lambda it will work, otherwise it dies with cryptic error messages :)

Reply:
As long as you have `.aws/credentials` in one of the appropriate places, or are in lambda it works like a charm, otherwise it dies with cryptic error messages...